src/viur/core/modules/user.py

```diff
@@ -341,6 +341,13 @@ def login(self, *, name: str | None = None, password: str | None = None, **kwarg
             skel["password"] = password  # will be hashed on serialize
             skel.toDB(update_relations=False)
 
+        return self.on_login(user_entry)
+
+    def on_login(self, user_entry: db.Entity):
```

Comment:
Can we move this in the abstract `UserAuthentication`? I think this can be useful for all methods.

Besides, can we not use action skels for this? You said yourself:

> I don't see any more hooks for actions in the VIUR system

_in https://github.com/viur-framework/viur-core/issues/972#issuecomment-1845301188_

Reply:
> Can we move this in the abstract `UserAuthentication`? I think this can be useful for all methods.

You're right, I thought about this as well.

> Besides, can we not use action skels for this? You said yourself:
> 
> > I don't see any more hooks for actions in the VIUR system
> 
> _in [#972 (comment)](https://github.com/viur-framework/viur-core/pull/972#issuecomment-1845301188)_

Well... let me clarify:

1. No, I don't want more explicit hooks for special cases, like a hook called `on_item_added_and_edited_and_deleted`
2. Yes, I want more hooks, but these should be universal and dynamical, like my proposed `on()`-hook which is called with event parameter on any action. You've stated that you want code completion for any VIUR modules, so we first have a to work out a solution which is both dynamic and fits all needs.
3. This is not an event hook, but I need a way to influence the behavior of the `UserPassword` primary authentication for custom behaviors - in this case an enforced password change before the login completes. Maybe a hook like this isn't the best way, but it works. I will update this PR soon and add the hook-function into the prototype.

Don't be angry if this comes across as a bit rough, I just want to have a general solution and not something that only works for one special case and not for everything else.

Reply:
> > Can we move this in the abstract `UserAuthentication`? I think this can be useful for all methods.
> 
> You're right, I thought about this as well.
> 

@sveneberth can we please merge it here and rewrite this PR for develop? Because in develop, the `UserPrimaryAuthentication` is also made available, which makes things easier. This is a feature needed by a current project, as it copies code from viur-core and extends the feature afterwards, as there is no hooking possible.

Reply:
Oh, we're here on the road to the main. Okay.